File: SPLL.m
function [Change,pst,st] = SPLL(W1,W2,PARAM)
%
% Detects a change in a multivariate data stream using the log-likelihood
% method
%
% [ Kuncheva L. I., Change detection in streaming multivariate data using
% likelihood detectors, IEEE Transactions on Knowledge and Data
% Engineering, 2013, 25(5), 1175-1180 (DOI: 10.1109/TKDE.2011.226). 
% http://pages.bangor.ac.uk/~mas00a/papers/lktkde11a.pdf ]
% 
% Input:
%     W1 = Data in Window 1:
%          matrix of size M1-by-n (M1 time points, n features)
%     W2 = Data in Window 2:
%          matrix of size M2-by-n (M2 time points, n features)
%     PARAM(1) = K - the number of clusters, default K = 3
%
% Output
%      Change = indicator: 0 - no change, 1 - change
%      pst = p-value
%      st = change statistic
%%
%========================================================================
% (c) Fox's Classification Toolbox                                  ^--^
% Last 28/04/2016 ------------------------------------------------  \oo/
%                                                                    \/
% Clustering within a window of size M1/M2 into K clusters
 
if nargin == 2 || isempty(PARAM)
    K = 3; % Predefined number of clusters
else
    K = PARAM;
end

n1 = size(W1,2);
n2 = size(W2,2);

if n1 == n2
    n = n1;
else
    error('The number of features must be the same for both windows')
end
 

s(1) = log_LL(W1,W2,K);
s(2) = log_LL(W2,W1,K);
 
st = max(s);
pst = min(chi2cdf(st,n),1-chi2cdf(st,n));
Change = pst < 0.05;
end
 
function st = log_LL(W1,W2,K)
 
 
[M1,n] = size(W1);
M2 = size(W2,1);

[labels,means] = kmeans(W1,K,'EmptyAction','singleton');
% uses stats toolbox
 
% Calculate the REFERENCE distribution from Window 1
 
SC = []; ClassPriors = zeros(1,K);
for k = 1:K
    ClassIndex = find(labels == k);
    ClassPriors(k) = numel(ClassIndex);
    if numel(ClassIndex) <= 1 % singleton cluster
        sc = zeros(n);
    else
        sc = cov(W1(labels == k,:));
    end
    SC = [SC;sc(:)']; %#ok<*AGROW>
end
ClassCount = ClassPriors;
ClassPriors = ClassPriors/M1;
 
scov = sum(SC.*repmat(ClassPriors',1,n^2),1);
 
% Assuming the same full covariance
% Pooled (weighted by the priors)
 
scov = reshape(scov,n,n);
z = diag(scov);
 
indexvarzero = z < eps; % identify features with very low variability
if ~sum(indexvarzero) % if there are no such features
    invscov = inv(scov); % invert the covariance matrix
else % if there are features with very low variance
    z(indexvarzero) = min(z(~indexvarzero)); % set their variance to the 
    % minimum variance among all other features
    invscov = diag(1./z); % invert the diagonal matrix only
end
 
% Calculate the MATCH from Window 2
for j = 1:M2
    xx = W2(j,:);
    
    for ii = 1:K
        if ClassCount(ii) > 0
            DistanceToMeans(ii) = (means(ii,:) - xx)...
                * invscov * (means(ii,:) - xx)';
        else
            DistanceToMeans(ii) = inf;
        end
    end
    
    LogLikelihoodTerm(j) = min(DistanceToMeans);
    
    % scale
end
 
st = mean(LogLikelihoodTerm); 
end

File: SPLL.py
##############################################################################
# Python/NumPy/Sklearn implementation of SPLL Change Detection Algorithm (Kuncheva, 2013)
# Based on Matlab code from https://github.com/LucyKuncheva/Change-detection
# This code was translated from Matlab to Python with minimal changes. Code is distributed under
# original Matlab code license (GPL-3.0)
# Author: Haidar Khan
# Date: 10/23/18
# Description: Implements the change point detection algorithm proposed by (Kuncheva, 2013) using a
# window based approach. 
# Usage: SPLL(W1,W2,K) tests if a change occurs between data windows W1 and W2 with K-Means clustering 
# parameter K

import numpy as np
from sklearn.cluster import KMeans
from scipy.stats import chi2
EPS = np.finfo(float).eps

def SPLL(W1, W2, K=3):
    n1 = W1.shape[1]
    n2 = W2.shape[1]

    assert n1==n2, "The number of features must be the same for W1 and W2"
    n = n1
    s1 = log_LL(W1,W2,K)
    s2 = log_LL(W2,W1,K)

    st = max(s1,s2)

    pst = min(chi2.cdf(st,n), 1-chi2.cdf(st,n))
    Change = float(pst < 0.05)
    return Change, pst, st

def log_LL(W1,W2,K):
    M1, n = W1.shape
    M2, _ = W2.shape

    kmeans = KMeans(n_clusters=K).fit(W1)

    labels = kmeans.predict(W1)
    means = kmeans.cluster_centers_

    SC = np.zeros((n*n, K))
    ClassPriors = np.zeros((K,1))
    for k in range(K):
        ClassIndex = np.nonzero(labels==k)[0]
        ClassPriors[k] = len(ClassIndex)
        if len(ClassIndex) <= 1:
            sc = np.zeros((n,n))
        else:
            sc = np.cov(W1[labels==k,:].T)
        SC[:,k] = sc.flatten()

    ClassCount = ClassPriors
    ClassPriors = ClassPriors/M1

    scov = np.sum(SC*np.tile(ClassPriors, (1,n*n)).T,1) 

    # Assuming the same full covariance
    # Pooled (weighted by the priors)
    scov = np.reshape(scov, (n,n))
    z = np.diag(scov)

    indexvarzero = z < EPS # define eps #identify features with very low variability
    if sum(indexvarzero) == 0:  #if no such features
        invscov = np.linalg.inv(scov)  # invert the covariance matrix
    else:   # otherwise
        z[indexvarzero] = np.min(z[not indexvarzero])   # set their variance to the minimum variance of other features
        invscov = np.diag(1/z) # invert the diagonal matrix only

    # Calculate the MATCH from Window 2
    LogLikelihoodTerm = np.zeros((M2))
    for j in range(M2):
        xx = W2[j,:]
        DistanceToMeans = np.zeros((K))
        for ii in range(K):
            if ClassCount[ii] > 0:
                DistanceToMeans[ii] = np.dot(np.dot(means[ii,:] - xx , invscov), means[ii,:]-xx)
            else:
                DistanceToMeans[ii] = np.inf
        LogLikelihoodTerm[j] = np.min(DistanceToMeans)
        # scale
    st = np.mean(LogLikelihoodTerm)
    return st

def test_SPLL():
    window_size = 100
    num_features = 10

    change1 = np.zeros((10))
    pvalue1 = np.zeros((10))
    spllstat1 = np.zeros((10))

    change2 = np.zeros((10))
    pvalue2 = np.zeros((10))
    spllstat2 = np.zeros((10))

    for i in range(10):
        W1 = np.random.randn(window_size, num_features)
        W2 = np.random.randn(window_size, num_features)
        W3 = np.random.randn(window_size, num_features)+2
        change1[i], pvalue1[i], spllstat1[i] = SPLL(W1, W2)
        change2[i], pvalue2[i], spllstat2[i] = SPLL(W1, W3)
    
        print "No change test: change: %d pval: %f stat: %f" %(change1[i], pvalue1[i], spllstat1[i])
        print "Change test:    change: %d pval: %f stat: %f" %(change2[i], pvalue2[i], spllstat2[i])

if __name__ == '__main__':
    test_SPLL()